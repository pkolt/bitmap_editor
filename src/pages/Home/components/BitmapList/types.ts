export enum SortDirection {
  NONE,
  ASC,
  DESC,
}
