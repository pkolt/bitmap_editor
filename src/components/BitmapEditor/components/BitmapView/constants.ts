// All sizes in pixels
export const SQUARE_SIZE = 10;
export const SQUARE_COLOR = '#000';
export const SEPARATOR_COLOR = '#777';
export const SEPARATOR_SIZE = 1;
export const BORDER_SIZE = 1;
export const BORDER_COLOR = '#ccc';
export const STEP_SIZE = SQUARE_SIZE + BORDER_SIZE;
export const AREA_POINT_COLOR = '#0d6efd';
export const AREA_OVERLAY_COLOR = '#0d6efd96';
