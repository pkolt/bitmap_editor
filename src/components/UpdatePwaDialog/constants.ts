export const PWA_REQUEST_UPDATE = 'PWA_REQUEST_UPDATE';
export const PWA_ACCEPT_UPDATE = 'PWA_ACCEPT_UPDATE';
