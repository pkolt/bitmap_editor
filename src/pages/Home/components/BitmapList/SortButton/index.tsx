import { SortValue } from '@/stores/settings';

interface Props {
  direction: SortValue;
  onClick: () => void;
  children: string;
}

const Config: Record<SortValue, { className: string }> = {
  [SortValue.NONE]: { className: 'bi-filter-circle' },
  [SortValue.DESC]: { className: 'bi-sort-down' },
  [SortValue.ASC]: { className: 'bi-sort-up-alt' },
};

export const SortButton = ({ direction, onClick, children }: Props) => {
  const { className } = Config[direction];
  return (
    <div className="d-flex align-items-center gap-2 user-select-none" role="button" onClick={onClick}>
      {children} <i className={className} />
    </div>
  );
};
