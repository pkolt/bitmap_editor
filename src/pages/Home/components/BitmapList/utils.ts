import { BitmapEntity } from '@/utils/bitmap/types';
import { SortValue } from '@/stores/settings';

const orderByFavorite = (list: BitmapEntity[]): BitmapEntity[] => {
  return list.toSorted((a, b) => Number(b.favorite) - Number(a.favorite));
};

const applyOrderDirection = (value: number, dir: SortValue): number => {
  return value * (dir === SortValue.DESC ? -1 : 1);
};

const orderByDate = (list: BitmapEntity[], dir: SortValue): BitmapEntity[] => {
  return list.toSorted((a, b) => applyOrderDirection(a.createdAt - b.createdAt, dir));
};

const orderByName = (list: BitmapEntity[], dir: SortValue): BitmapEntity[] => {
  return list.toSorted((a, b) => applyOrderDirection(b.name.localeCompare(a.name), dir));
};

export const orderBitmaps = (bitmaps: BitmapEntity[], nameDir: SortValue, dateDir: SortValue): BitmapEntity[] => {
  let orderedList: BitmapEntity[] = bitmaps;
  if (dateDir !== SortValue.NONE) {
    orderedList = orderByDate(orderedList, dateDir);
  }
  if (nameDir !== SortValue.NONE) {
    orderedList = orderByName(orderedList, nameDir);
  }
  return orderByFavorite(orderedList);
};

const ORDERED_VALUES: SortValue[] = [SortValue.NONE, SortValue.DESC, SortValue.ASC] as const;

export const getNextSortValue = (curSortValue: SortValue): SortValue => {
  const index = ORDERED_VALUES.indexOf(curSortValue);
  const nextIndex = index < ORDERED_VALUES.length - 1 ? index + 1 : 0;
  const nextValue = ORDERED_VALUES[nextIndex];
  return nextValue;
};
