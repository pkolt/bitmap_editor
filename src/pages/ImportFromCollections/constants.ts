export const ICONS_DATA_URL = `${import.meta.env.BASE_URL}images/bootstrap-icons/data.json`;
export const ICONS_DIR_URL = `${import.meta.env.BASE_URL}images/bootstrap-icons/icons/`;
