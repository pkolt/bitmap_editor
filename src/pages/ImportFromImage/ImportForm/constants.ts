import { FormValues } from './types';

export const defaultValues: FormValues = {
  files: null,
  top: 0,
  left: 0,
  width: 128,
  height: 64,
  threshold: 100,
  name: '',
  invertColor: true,
};
