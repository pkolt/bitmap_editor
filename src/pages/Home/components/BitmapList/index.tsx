import { useBitmapsStore } from '@/stores/bitmaps';
import { useMemo } from 'react';
import { OpenDialogFn } from '../../types';
import { BitmapItem } from '../BitmapItem';
import { useTranslation } from 'react-i18next';
import { SortDirection } from './types';
import { SortButton } from './SortButton';
import { useSortButton } from './SortButton/hooks';
import { orderBitmaps } from './utils';

interface BitmapListProps {
  openDialog: OpenDialogFn;
}

export const BitmapList = ({ openDialog }: BitmapListProps): JSX.Element | null => {
  const { t } = useTranslation();
  const { bitmaps } = useBitmapsStore();
  const nameSortBtn = useSortButton(SortDirection.NONE);
  const dateSortBtn = useSortButton(SortDirection.DESC);
  const bitmapIds: string[] = useMemo(() => {
    return orderBitmaps(bitmaps, nameSortBtn.direction, dateSortBtn.direction).map((it) => it.id);
  }, [bitmaps, dateSortBtn.direction, nameSortBtn.direction]);

  if (bitmapIds.length === 0) {
    return null;
  }

  return (
    <div className="mb-3" data-testid="bitmap-list">
      <h2 className="text-center mb-3">{t('Open')}</h2>
      <table className="table table-borderless mb-3">
        <thead>
          <tr>
            <th></th>
            <th>
              <SortButton direction={nameSortBtn.direction} onClick={nameSortBtn.toggle}>
                {t('Name')}
              </SortButton>
            </th>
            <th className="text-center">{t('Size')}</th>
            <th>
              <SortButton direction={dateSortBtn.direction} onClick={dateSortBtn.toggle}>
                {t('Created')}
              </SortButton>
            </th>
            <th colSpan={3}>{t('Actions')}</th>
          </tr>
        </thead>
        <tbody>
          {bitmapIds.map((id) => (
            <BitmapItem key={id} bitmapId={id} openDialog={openDialog} />
          ))}
        </tbody>
      </table>
      <h5 className="d-flex gap-3 align-items-center">
        <hr className="flex-grow-1" />
        {t('Or')}
        <hr className="flex-grow-1" />
      </h5>
    </div>
  );
};
