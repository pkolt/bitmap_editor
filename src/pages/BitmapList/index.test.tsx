import { test, renderPage, expect } from '@/test-utils';
import BitmapList from './';
import { PageUrl } from '@/constants/urls';

test('show empty', async () => {
  const { container, location } = renderPage(<BitmapList />, { route: { path: PageUrl.BitmapList } });
  expect(location().pathname).toEqual(PageUrl.BitmapList);
  expect(container).matchSnapshot();
});
