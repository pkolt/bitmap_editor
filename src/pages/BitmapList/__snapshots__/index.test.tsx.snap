// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`show empty 1`] = `
<div>
  <div
    class="container d-flex flex-column gap-4 min-vh-100"
  >
    <div
      class="navbar bg-body-tertiary rounded-bottom-2"
    >
      <div
        class="container-fluid"
      >
        <a
          class="navbar-brand d-flex gap-2 align-items-center"
          href="/bitmap_editor/"
        >
          <i
            class="bi bi-border-all display-6"
          />
          <div
            class="d-flex flex-column"
          >
            <div
              class="lh-sm"
            >
              Bitmap Editor
            </div>
            <div
              class="h6 text-body-tertiary mb-0"
            >
              Create bitmap image for OLED display
            </div>
          </div>
        </a>
        <a
          class="h3 m-0"
          href="https://github.com/pkolt/bitmap_editor"
          target="_blank"
        >
          <i
            class="bi bi-github"
          />
        </a>
      </div>
    </div>
    <main
      class="d-flex flex-column flex-grow-1 justify-content-center align-items-center"
    >
      <div
        class="d-flex flex-column gap-3"
      >
        <a
          class="btn btn-primary btn-lg"
          href="/bitmap_editor/create_bitmap/"
        >
          Create new bitmap
        </a>
        <a
          class="btn btn-primary btn-lg"
          href="/bitmap_editor/import_from_image/"
        >
          Import from image
        </a>
        <a
          class="btn btn-primary btn-lg"
          href="/bitmap_editor/import_from_json/"
        >
          Import from JSON
        </a>
      </div>
    </main>
    <footer
      class="container-fluid p-3 bg-light text-center rounded-top-2"
    >
      Created by 
      <a
        href="https://github.com/pkolt"
      >
        Pavel Koltyshev
      </a>
       Â© 2023-
      2024
    </footer>
  </div>
</div>
`;
